internal/service/subscription.go

```diff
@@ -371,14 +371,37 @@ func (s *subscriptionService) handleCreditGrants(
 	for _, grant := range creditGrants {
 		// Calculate expiry date if needed
 		var expiryDate *time.Time
-		if grant.ExpireInDays != nil && *grant.ExpireInDays > 0 {
-			expiry := subscription.StartDate.AddDate(0, 0, *grant.ExpireInDays)
-			expiryDate = &expiry
+
+		if grant.ExpirationType == types.CreditGrantExpiryTypeNever {
+			expiryDate = nil
+		}
+
+		if grant.ExpirationType == types.CreditGrantExpiryTypeDuration {
+			if grant.ExpirationDurationUnit != nil && grant.ExpirationDuration != nil && *grant.ExpirationDuration > 0 {
+				switch *grant.ExpirationDurationUnit {
+				case types.CreditGrantExpiryDurationUnitDays:
+					expiry := subscription.StartDate.AddDate(0, 0, *grant.ExpirationDuration)
+					expiryDate = &expiry
+				case types.CreditGrantExpiryDurationUnitWeeks:
+					expiry := subscription.StartDate.AddDate(0, 0, *grant.ExpirationDuration*7)
+					expiryDate = &expiry
+				case types.CreditGrantExpiryDurationUnitMonths:
+					expiry := subscription.StartDate.AddDate(0, *grant.ExpirationDuration, 0)
+					expiryDate = &expiry
+				case types.CreditGrantExpiryDurationUnitYears:
+					expiry := subscription.StartDate.AddDate(*grant.ExpirationDuration, 0, 0)
+					expiryDate = &expiry
+				}
+			}
+		}
+
+		if grant.ExpirationType == types.CreditGrantExpiryTypeBillingCycle {
+			expiryDate = &subscription.CurrentPeriodEnd
```

Comment:
TODO: separate this out as a helper function and use switch cases or if else block